In [34]:
import pandas as pd
import numpy as np
from imblearn.over_sampling import RandomOverSampler
from sklearn.datasets import make_classification
from datetime import datetime

from sklearn.model_selection import train_test_split

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)


In [35]:
ml_dataframe_df = pd.read_csv("data/ml_dataframe.csv")


In [36]:
ml_dataframe_df.head()

driverId  raceId         GP  year_x driverRef_x  Full Wet  Hard New   
0       1.0   860.0  Australia    2012    hamilton         0         0  \
1       1.0   880.0  Australia    2013    hamilton         0         0   
2       1.0   926.0  Australia    2015    hamilton         0         0   
3       1.0   948.0  Australia    2016    hamilton         0         0   
4       1.0   969.0  Australia    2017    hamilton         0         0   

   Hard Used  Hyper Soft  Intermediate New  Intermediate Used  Medium New   
0          0           0                 0                  0          41  \
1          0           0                 0                  0          45   
2          0           0                 0                  0          33   
3          0           0                 0                  0          41   
4          0           0                 0                  0           0   

   Medium Used  Soft New  Soft Used  Supersoft New  Supersoft Used   
0            0         0         17              0               0  \
1            0         0          0              0              13   
2            0         0         25              0               0   
3            0         0          0              0              16   
4            0        40          0              0               0   

   Ultrasoft New  Ultrasoft Used  Wet New  circuitId        date   
0              0               0        0          1  2012-03-18  \
1              0               0        0          1  2013-03-17   
2              0               0        0          1  2015-03-15   
3              0               0        0          1  2016-03-20   
4              0              17        0          1  2017-03-26   

   weather_cloudy  weather_cold  weather_dry  weather_warm  weather_wet   
0           False         False        False          True        False  \
1           False         False        False          True        False   
2            True         False        False         False        False   
3            True         False        False         False        False   
4            True         False        False         False        False   

    circuitRef location_y    country  resultId  constructorId  grid   
0  albert_park  Melbourne  Australia     21234              1     1  \
1  albert_park  Melbourne  Australia     21716            131     3   
2  albert_park  Melbourne  Australia     22538            131     1   
3  albert_park  Melbourne  Australia     22918            131     1   
4  albert_park  Melbourne  Australia     23380            131     1   

  positionText  positionOrder  points  laps fastestLap  rank fastestLapTime   
0            3              3    15.0    58         57     4       1:29.538  \
1            5              5    10.0    58         45     5       1:29.759   
2            1              1    25.0    58         50     1       1:30.945   
3            2              2    18.0    57         48     4       1:30.646   
4            2              2    18.0    57         44     6       1:27.033   

  fastestLapSpeed  statusId nationality  pit_stop_1_lap_x pit_stop_1_time_x   
0         213.214         1     British              17.0          17:30:12  \
1         212.689         1     British              13.0          17:23:56   
2         209.915         1     British              25.0          16:44:32   
3         210.608         1     British              16.0          16:32:16   
4         219.351         1     British              17.0          16:31:40   

  pit_stop_1_duration_x  pit_stop_1_milliseconds_x  pit_stop_2_lap_x   
0                22.862                    22862.0              36.0  \
1                22.155                    22155.0              31.0   
2                22.295                    22295.0               0.0   
3                21.940                    21940.0              18.0   
4                21.709                    21709.0               0.0   

  pi

In [37]:
ml_dataframe_df.columns

Index(['driverId', 'raceId', 'GP', 'year_x', 'driverRef_x', 'Full Wet',
       'Hard New', 'Hard Used', 'Hyper Soft', 'Intermediate New',
       'Intermediate Used', 'Medium New', 'Medium Used', 'Soft New',
       'Soft Used', 'Supersoft New', 'Supersoft Used', 'Ultrasoft New',
       'Ultrasoft Used', 'Wet New', 'circuitId', 'date', 'weather_cloudy',
       'weather_cold', 'weather_dry', 'weather_warm', 'weather_wet',
       'circuitRef', 'location_y', 'country', 'resultId', 'constructorId',
       'grid', 'positionText', 'positionOrder', 'points', 'laps', 'fastestLap',
       'rank', 'fastestLapTime', 'fastestLapSpeed', 'statusId', 'nationality',
       'pit_stop_1_lap_x', 'pit_stop_1_time_x', 'pit_stop_1_duration_x',
       'pit_stop_1_milliseconds_x', 'pit_stop_2_lap_x', 'pit_stop_2_time_x',
       'pit_stop_2_duration_x', 'pit_stop_2_milliseconds_x',
       'pit_stop_3_lap_x', 'pit_stop_3_time_x', 'pit_stop_3_duration_x',
       'pit_stop_3_milliseconds_x', 'pit_stop_4_lap_x', 'pi

In [38]:
columns_to_keep = ['driverId','raceId', 'grid','positionOrder', 'year_x','Full Wet',
       'Hard New', 'Hard Used', 'Hyper Soft', 'Intermediate New',
       'Intermediate Used', 'Medium New', 'Medium Used', 'Soft New',
       'Soft Used', 'Supersoft New', 'Supersoft Used', 'Ultrasoft New',
       'Ultrasoft Used', 'Wet New', 'circuitId', 'weather_cloudy',
       'weather_cold', 'weather_dry', 'weather_warm', 'weather_wet', 'pit_stop_1_lap_x',
       'pit_stop_1_milliseconds_x', 'pit_stop_2_lap_x', 
        'pit_stop_2_milliseconds_x',
       'pit_stop_3_lap_x', 
       'pit_stop_3_milliseconds_x', 'pit_stop_4_lap_x',
       'pit_stop_4_milliseconds_x',
       'pit_stop_5_lap_x', 
       'pit_stop_5_milliseconds_x', 'pit_stop_6_lap_x',
       'pit_stop_6_milliseconds_x']

new_merged_df = ml_dataframe_df[columns_to_keep].copy()

print(new_merged_df)

      driverId  raceId  grid  positionOrder  year_x  Full Wet  Hard New   
0          1.0   860.0     1              3    2012         0         0  \
1          1.0   880.0     3              5    2013         0         0   
2          1.0   926.0     1              1    2015         0         0   
3          1.0   948.0     1              2    2016         0         0   
4          1.0   969.0     1              2    2017         0         0   
5          1.0   989.0     1              2    2018         0         0   
6          1.0  1010.0     1              2    2019         0         0   
7          1.0  1076.0     5              4    2022         0        36   
8          1.0   861.0     1              3    2012         0         0   
9          1.0   881.0     4              3    2013         0        11   
10         1.0   901.0     1              1    2014         0         5   
11         1.0   927.0     1              2    2015         0        38   
12         1.0   963.0   

In [39]:
new_merged_df.head()

driverId  raceId  grid  positionOrder  year_x  Full Wet  Hard New   
0       1.0   860.0     1              3    2012         0         0  \
1       1.0   880.0     3              5    2013         0         0   
2       1.0   926.0     1              1    2015         0         0   
3       1.0   948.0     1              2    2016         0         0   
4       1.0   969.0     1              2    2017         0         0   

   Hard Used  Hyper Soft  Intermediate New  Intermediate Used  Medium New   
0          0           0                 0                  0          41  \
1          0           0                 0                  0          45   
2          0           0                 0                  0          33   
3          0           0                 0                  0          41   
4          0           0                 0                  0           0   

   Medium Used  Soft New  Soft Used  Supersoft New  Supersoft Used   
0            0         0         17              0               0  \
1            0         0          0              0              13   
2            0         0         25              0               0   
3            0         0          0              0              16   
4            0        40          0              0               0   

   Ultrasoft New  Ultrasoft Used  Wet New  circuitId  weather_cloudy   
0              0               0        0          1           False  \
1              0               0        0          1           False   
2              0               0        0          1            True   
3              0               0        0          1            True   
4              0              17        0          1            True   

   weather_cold  weather_dry  weather_warm  weather_wet  pit_stop_1_lap_x   
0         False        False          True        False              17.0  \
1         False        False          True        False              13.0   
2         False        False         False        False              25.0   
3         False        False         False        False              16.0   
4         False        False         False        False              17.0   

   pit_stop_1_milliseconds_x  pit_stop_2_lap_x  pit_stop_2_milliseconds_x   
0                    22862.0              36.0                    23464.0  \
1                    22155.0              31.0                    22015.0   
2                    22295.0               0.0                        0.0   
3                    21940.0              18.0                  1090087.0   
4                    21709.0               0.0                        0.0   

   pit_stop_3_lap_x  pit_stop_3_milliseconds_x  pit_stop_4_lap_x   
0               0.0                        0.0               0.0  \
1              42.0                    21778.0               0.0   
2               0.0                        0.0               0.0   
3               0.0                        0.0               0.0   
4               0.0                        0.0               0.0   

   pit_stop_4_milliseconds_x  pit_stop_5_lap_x  pit_stop_5_milliseconds_x   
0                        0.0               0.0                        0.0  \
1                        0.0               0.0                        0.0   
2                        0.0               0.0                        0.0   
3                        0.0               0.0                        0.0   
4                        0.0               0.0                        0.0   

   pit_stop_6_lap_x  pit_stop_6_milliseconds_x  
0               0.0                        0.0  
1               0.0                        0.0  
2               0.0                        0.0  
3               0.0                        0.0  
4               0.0                        0.0

#### Oversampling

- Separate the features (X) and the target variable (y). In your case, it seems like you want to oversample based on some specific columns (e.g., 'positionOrder'):

In [40]:
# Assuming 'positionOrder' is your target variable, and you want to oversample based on it
X = new_merged_df.drop('positionOrder', axis=1)  # Features
y = new_merged_df['positionOrder']  # Target variable


- Split dataset into training and testing subsets:


In [41]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)


- Create an instance of the RandomOverSampler:


In [42]:
oversampler = RandomOverSampler(random_state=42)


- Use the oversampler to fit and transform your training data to oversample the minority class:


In [43]:
X_train_resampled, y_train_resampled = oversampler.fit_resample(X_train, y_train)


In [44]:
# Combine the oversampled training features and target variable into a new DataFrame
oversampled_train_df = pd.DataFrame(X_train_resampled, columns=X.columns)
oversampled_train_df['positionOrder'] = y_train_resampled


In [45]:
oversampled_train_df.head(10)

driverId  raceId  grid  year_x  Full Wet  Hard New  Hard Used  Hyper Soft   
0     817.0   965.0     3    2016         0         0          0           0  \
1     820.0   883.0    21    2013         0        27          9           0   
2     841.0   969.0    16    2017         0         0          0           0   
3     832.0  1084.0     3    2022         0        29          0           0   
4     852.0  1080.0    11    2022         0         8          0           0   
5     826.0  1029.0    16    2019         0         0          0           0   
6     828.0   951.0    22    2016         0         0          0           0   
7     847.0  1024.0    18    2019         0        33          0           0   
8     849.0  1084.0    17    2022         0        36          0           0   
9     815.0   895.0    11    2013         0        30          0           0   

   Intermediate New  Intermediate Used  Medium New  Medium Used  Soft New   
0                 0                  0          31            0        17  \
1                 0                  0          10           10         0   
2                 0                  0           0            0        15   
3                 0                  0          27            0         0   
4                15                  0          27            0         7   
5                 0                  0          23            0        48   
6                 0                  0           0            0         1   
7                 0                  0           1            0         0   
8                 0                  0          12            0         0   
9                 0                  0          12           10         0   

   Soft Used  Supersoft New  Supersoft Used  Ultrasoft New  Ultrasoft Used   
0          0              0               8              0               0  \
1          0              0               0              0               0   
2          0             40               0              0               0   
3          0              0               0              0               0   
4          0              0               0              0               0   
5          0              0               0              0               0   
6          0             51               0              0               0   
7          0              0               0              0               0   
8          0              0               0              0               0   
9          0              0               0              0               0   

   Wet New  circuitId  weather_cloudy  weather_cold  weather_dry   
0        0         69           False         False        False  \
1        0          3           False         False        False   
2        0          1            True         False        False   
3        0         70            True         False        False   
4        6          6           False         False         True   
5        0         18           False         False        False   
6        0         71            True         False        False   
7        0         15           False         False        False   
8        0         70            True         False        False   
9        0         22           False         False        False   

   weather_warm  weather_wet  pit_stop_1_lap_x  pit_stop_1_milliseconds_x   
0          True        False               8.0                    23956.0  \
1          True        False              10.0                    22745.0   
2         False        False              15.0                    28591.0   
3         False        False              27.0                    21212.0   
4         False        False               6.0                    25068.0   
5          True        False              23.0                    22759.0   
6         False        False               1.0                    30529.0   
7         False        False               1.0    

In [46]:
oversampled_train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2184 entries, 0 to 2183
Data columns (total 38 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   driverId                   2184 non-null   float64
 1   raceId                     2184 non-null   float64
 2   grid                       2184 non-null   int64  
 3   year_x                     2184 non-null   int64  
 4   Full Wet                   2184 non-null   int64  
 5   Hard New                   2184 non-null   int64  
 6   Hard Used                  2184 non-null   int64  
 7   Hyper Soft                 2184 non-null   int64  
 8   Intermediate New           2184 non-null   int64  
 9   Intermediate Used          2184 non-null   int64  
 10  Medium New                 2184 non-null   int64  
 11  Medium Used                2184 non-null   int64  
 12  Soft New                   2184 non-null   int64  
 13  Soft Used                  2184 non-null   int64

In [50]:


# Assuming you have a DataFrame called oversampled_train_df with a 'positionOrder' column
# Create a new column 'Class' based on the condition
oversampled_train_df['Class'] = oversampled_train_df['positionOrder'].apply(lambda x: 1 if x in [1, 2, 3] else 0)

# Split the DataFrame into features (X) and target (y)
X = oversampled_train_df.drop(columns=['positionOrder', 'Class'])  # Assuming other columns are features
y = oversampled_train_df['Class']

# Instantiate the RandomOverSampler
oversampler = RandomOverSampler(random_state=42)

# Fit and transform the data to oversample the minority class
X_resampled, y_resampled = oversampler.fit_resample(X, y)

# Create a new balanced DataFrame
balanced_df = pd.concat([X_resampled, y_resampled], axis=1)

# Check the new class distribution in the balanced DataFrame
balanced_class_distribution = balanced_df['Class'].value_counts()
print("Balanced Class Distribution:")
print(balanced_class_distribution)

Balanced Class Distribution:
Class
1    1911
0    1911
Name: count, dtype: int64


In [51]:
balanced_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3822 entries, 0 to 3821
Data columns (total 38 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   driverId                   3822 non-null   float64
 1   raceId                     3822 non-null   float64
 2   grid                       3822 non-null   int64  
 3   year_x                     3822 non-null   int64  
 4   Full Wet                   3822 non-null   int64  
 5   Hard New                   3822 non-null   int64  
 6   Hard Used                  3822 non-null   int64  
 7   Hyper Soft                 3822 non-null   int64  
 8   Intermediate New           3822 non-null   int64  
 9   Intermediate Used          3822 non-null   int64  
 10  Medium New                 3822 non-null   int64  
 11  Medium Used                3822 non-null   int64  
 12  Soft New                   3822 non-null   int64  
 13  Soft Used                  3822 non-null   int64

In [52]:
balanced_df.head()

driverId  raceId  grid  year_x  Full Wet  Hard New  Hard Used  Hyper Soft   
0     817.0   965.0     3    2016         0         0          0           0  \
1     820.0   883.0    21    2013         0        27          9           0   
2     841.0   969.0    16    2017         0         0          0           0   
3     832.0  1084.0     3    2022         0        29          0           0   
4     852.0  1080.0    11    2022         0         8          0           0   

   Intermediate New  Intermediate Used  Medium New  Medium Used  Soft New   
0                 0                  0          31            0        17  \
1                 0                  0          10           10         0   
2                 0                  0           0            0        15   
3                 0                  0          27            0         0   
4                15                  0          27            0         7   

   Soft Used  Supersoft New  Supersoft Used  Ultrasoft New  Ultrasoft Used   
0          0              0               8              0               0  \
1          0              0               0              0               0   
2          0             40               0              0               0   
3          0              0               0              0               0   
4          0              0               0              0               0   

   Wet New  circuitId  weather_cloudy  weather_cold  weather_dry   
0        0         69           False         False        False  \
1        0          3           False         False        False   
2        0          1            True         False        False   
3        0         70            True         False        False   
4        6          6           False         False         True   

   weather_warm  weather_wet  pit_stop_1_lap_x  pit_stop_1_milliseconds_x   
0          True        False               8.0                    23956.0  \
1          True        False              10.0                    22745.0   
2         False        False              15.0                    28591.0   
3         False        False              27.0                    21212.0   
4         False        False               6.0                    25068.0   

   pit_stop_2_lap_x  pit_stop_2_milliseconds_x  pit_stop_3_lap_x   
0              25.0                    23534.0               0.0  \
1              23.0                    24141.0              37.0   
2               0.0                        0.0               0.0   
3              50.0                    21492.0               0.0   
4              21.0                    26021.0              29.0   

   pit_stop_3_milliseconds_x  pit_stop_4_lap_x  pit_stop_4_milliseconds_x   
0                        0.0               0.0                        0.0  \
1                    26004.0              46.0                    22552.0   
2                        0.0               0.0                        0.0   
3                        0.0               0.0                        0.0   
4                  1174444.0              56.0                    25266.0   

   pit_stop_5_lap_x  pit_stop_5_milliseconds_x  pit_stop_6_lap_x   
0               0.0                        0.0               0.0  \
1               0.0                        0.0               0.0   
2               0.0                        0.0               0.0   
3               0.0                        0.0               0.0   
4               0.0                        0.0               0.0   

   pit_stop_6_milliseconds_x  Class  
0                        0.0      1  
1                        0.0      0  
2                        0.0      0  
3                        0.0      0  
4                        0.0      0

- So in the last column of the above dataframe we have a column called Class.
- Class 1: the top 3 podium positions
- Class 0: positions from 4 to 20

In [53]:
# Assuming you have a DataFrame called oversampled_train_df with a 'positionOrder' column and a 'year_x' column in datetime format

# Define the training and testing year_x ranges
training_start_date = 2012
training_end_date = 2018
testing_start_date = 2019
testing_end_date = 2022

# Filter the DataFrame for training and testing data
training_df = oversampled_train_df[(oversampled_train_df['year_x'] >= training_start_date) & (oversampled_train_df['year_x'] <= training_end_date)]
testing_df = oversampled_train_df[(oversampled_train_df['year_x'] >= testing_start_date) & (oversampled_train_df['year_x'] <= testing_end_date)]

# Create a new column 'Class' based on the condition for both training and testing DataFrames
training_df['Class'] = training_df['positionOrder'].apply(lambda x: 1 if x in [1, 2, 3] else 0)
testing_df['Class'] = testing_df['positionOrder'].apply(lambda x: 1 if x in [1, 2, 3] else 0)

# Split the training and testing DataFrames into features (X) and target (y)
X_train = training_df.drop(columns=['positionOrder', 'Class'])  # Assuming other columns are features
y_train = training_df['Class']

X_test = testing_df.drop(columns=['positionOrder', 'Class'])  # Assuming other columns are features
y_test = testing_df['Class']

# Instantiate the RandomOverSampler for training data
oversampler = RandomOverSampler(random_state=42)

# Fit and transform the training data to oversample the minority class
X_train_resampled, y_train_resampled = oversampler.fit_resample(X_train, y_train)

# Create a new balanced training DataFrame
balanced_train_df = pd.concat([X_train_resampled, y_train_resampled], axis=1)

# Check the new class distribution in the balanced training DataFrame
balanced_train_class_distribution = balanced_train_df['Class'].value_counts()
print("Balanced Training Class Distribution:")
print(balanced_train_class_distribution)

# The testing data (X_test and y_test) remains unmodified for evaluation


Balanced Training Class Distribution:
Class
1    1357
0    1357
Name: count, dtype: int64


/var/folders/g4/vm6jsw_s0hlcmn6jslrlcxm80000gp/T/ipykernel_33140/465968541.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  training_df['Class'] = training_df['positionOrder'].apply(lambda x: 1 if x in [1, 2, 3] else 0)
/var/folders/g4/vm6jsw_s0hlcmn6jslrlcxm80000gp/T/ipykernel_33140/465968541.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  testing_df['Class'] = testing_df['positionOrder'].apply(lambda x: 1 if x in [1, 2, 3] else 0)


#### Train our model

##### KNN

In [54]:
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report

In [55]:
# Step 1: Data Preprocessing
X = balanced_df.drop(columns=['Class'])  # Features
y = balanced_df['Class']  # Target variable

# Step 2: Data Splitting
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Step 3: Model Selection - KNN
k = 3  
model = KNeighborsClassifier(n_neighbors=k)

# Step 4: Feature Scaling (Important for KNN)
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Step 5: Model Training
model.fit(X_train_scaled, y_train)

# Step 6: Model Evaluation
y_pred = model.predict(X_test_scaled)
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy}')
print(classification_report(y_test, y_pred))


Accuracy: 0.9267973856209151
              precision    recall  f1-score   support

           0       0.99      0.86      0.92       398
           1       0.87      0.99      0.93       367

    accuracy                           0.93       765
   macro avg       0.93      0.93      0.93       765
weighted avg       0.94      0.93      0.93       765



#### Grid Search for Hyperparameter Tuning


In [56]:
from sklearn.model_selection import train_test_split, GridSearchCV


# Step 1: Data Preprocessing
X = balanced_df.drop(columns=['Class'])  # Features
y = balanced_df['Class']  # Target variable

# Step 2: Data Splitting
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Step 3: Define the parameter grid for grid search
param_grid = {
    'n_neighbors': [3, 5, 7, 9, 11],  # You can define more values to try
}

# Step 4: Create the KNN classifier
knn = KNeighborsClassifier()

# Step 5: Perform grid search with cross-validation
grid_search = GridSearchCV(knn, param_grid, cv=5)  # You can adjust the number of folds (cv) as needed

# Step 6: Fit the grid search to the data
grid_search.fit(X_train, y_train)

# Step 7: Get the results for all tested values of k
results = grid_search.cv_results_
for mean_score, params in zip(results['mean_test_score'], results['params']):
    print(f"Mean Accuracy: {mean_score:.4f} for params: {params}")

# Step 8: Get the best hyperparameters
best_k = grid_search.best_params_['n_neighbors']

# Step 9: Use the best hyperparameters to create the final KNN model
best_knn = KNeighborsClassifier(n_neighbors=best_k)
best_knn.fit(X_train, y_train)

# Step 10: Evaluate the final model
y_pred = best_knn.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(f'Best k: {best_k}')
print(f'Accuracy: {accuracy}')
print(classification_report(y_test, y_pred))


Mean Accuracy: 0.8829 for params: {'n_neighbors': 3}
Mean Accuracy: 0.8335 for params: {'n_neighbors': 5}
Mean Accuracy: 0.7949 for params: {'n_neighbors': 7}
Mean Accuracy: 0.7642 for params: {'n_neighbors': 9}
Mean Accuracy: 0.7416 for params: {'n_neighbors': 11}
Best k: 3
Accuracy: 0.8993464052287582
              precision    recall  f1-score   support

           0       0.98      0.82      0.89       398
           1       0.84      0.98      0.90       367

    accuracy                           0.90       765
   macro avg       0.91      0.90      0.90       765
weighted avg       0.91      0.90      0.90       765



The k=3 has high accuracy and balanced performance in terms of precision, recall, and F1-score for both classes (0 and 1). This indicates that the model is effective at identifying drivers who finish in the top 3 positions (class 1) and those who don't (class 0).

#### Confusion Matrix

In [57]:
from sklearn.metrics import confusion_matrix
confusion = confusion_matrix(y_test, y_pred)
print(confusion)

[[327  71]
 [  6 361]]


- True Positives (TP): 361
- True Negatives (TN): 327
- False Positives (FP): 71
- False Negatives (FN): 6

In [58]:
import pandas as pd
from sklearn.model_selection import cross_val_score, train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report

# Load your dataset into balanced_df

# Step 1: Data Preprocessing
X = balanced_df.drop(columns=['Class'])  # Features
y = balanced_df['Class']  # Target variable

# Step 2: Split the data into a training set and a testing set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Step 3: Create the KNN classifier
knn = KNeighborsClassifier(n_neighbors=5)  # You can set the value of k

# Step 4: Perform Cross-Validation
num_folds = 5  # You can adjust the number of folds as needed
scoring_metric = 'accuracy'  # You can choose a different scoring metric (e.g., 'f1', 'precision', 'recall')

# Use cross_val_score to perform cross-validation and obtain scores
cv_scores = cross_val_score(knn, X_train, y_train, cv=num_folds, scoring=scoring_metric)

# Print the cross-validation scores for each fold
for fold, score in enumerate(cv_scores, start=1):
    print(f'Fold {fold}: {score:.4f}')

# Calculate and print the mean and standard deviation of the scores
mean_score = cv_scores.mean()
std_score = cv_scores.std()
print(f'Mean {scoring_metric}: {mean_score:.4f}')
print(f'Standard Deviation: {std_score:.4f}')

# Step 5: Fit and evaluate the final model on the test set
knn.fit(X_train, y_train)
y_pred = knn.predict(X_test)
test_accuracy = accuracy_score(y_test, y_pred)
print(f'Test Accuracy: {test_accuracy:.4f}')
print(classification_report(y_test, y_pred))


Fold 1: 0.8317
Fold 2: 0.8317
Fold 3: 0.8527
Fold 4: 0.8511
Fold 5: 0.8003
Mean accuracy: 0.8335
Standard Deviation: 0.0189
Test Accuracy: 0.8575
              precision    recall  f1-score   support

           0       0.95      0.77      0.85       398
           1       0.79      0.96      0.87       367

    accuracy                           0.86       765
   macro avg       0.87      0.86      0.86       765
weighted avg       0.87      0.86      0.86       765



##### Logistic Regression

In [59]:
from sklearn.linear_model import LogisticRegression

# Step 1: Data Preprocessing
X = balanced_df.drop(columns=['Class'])  # Features
y = balanced_df['Class']  # Target variable

# Step 2: Data Splitting
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Step 3: Create the Logistic Regression model
logistic_model = LogisticRegression(random_state=42)

# Step 4: Train the model on the training data
logistic_model.fit(X_train, y_train)

# Step 5: Make predictions on the test data
y_pred = logistic_model.predict(X_test)

# Step 6: Evaluate the model's performance
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy}')
print(classification_report(y_test, y_pred))


Accuracy: 0.6627450980392157
              precision    recall  f1-score   support

           0       0.68      0.66      0.67       398
           1       0.64      0.67      0.66       367

    accuracy                           0.66       765
   macro avg       0.66      0.66      0.66       765
weighted avg       0.66      0.66      0.66       765



/Users/Ana/neuefische/Formula-1-Project/.venv/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


The classification report shows that the Logistic Regression model achieves a balanced performance with similar precision, recall, and F1-score for both classes (0 and 1). The accuracy of 66.27% indicates that the model is moderately successful at classifying drivers who finish in the top 3 positions.

Comparing this result with the KNN model (where the best k was 3), it appears that the KNN model had a higher accuracy of approximately 89.93%. Therefore, based on accuracy alone, the KNN model outperforms the Logistic Regression model. 

 Let's perform hyperparameter tuning and apply regularization to our Logistic Regression model using scikit-learn to improve our results.

In [60]:

# Step 1: Data Preprocessing
X = balanced_df.drop(columns=['Class'])  # Features
y = balanced_df['Class']  # Target variable

# Step 2: Data Splitting
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Step 3: Create the Logistic Regression model
logistic_model = LogisticRegression(random_state=42)

# Step 4: Define a hyperparameter grid for tuning
param_grid = {
    'C': [0.001, 0.01, 0.1, 1, 10, 100],  # Different values of regularization strength (C)
    'penalty': ['l1', 'l2'],  # Regularization type (L1 or L2)
}

# Step 5: Create a GridSearchCV object for hyperparameter tuning
grid_search = GridSearchCV(logistic_model, param_grid, cv=5)  # You can adjust the number of folds (cv) as needed

# Step 6: Fit the grid search to the data
grid_search.fit(X_train, y_train)

# Step 7: Get the best hyperparameters
best_params = grid_search.best_params_

# Step 8: Use the best hyperparameters to create the final Logistic Regression model
best_logistic_model = LogisticRegression(C=best_params['C'], penalty=best_params['penalty'], random_state=42)

# Step 9: Train the final model on the training data
best_logistic_model.fit(X_train, y_train)

# Step 10: Make predictions on the test data
y_pred = best_logistic_model.predict(X_test)

# Step 11: Evaluate the final model's performance
accuracy = accuracy_score(y_test, y_pred)
print(f'Best hyperparameters: {best_params}')
print(f'Accuracy: {accuracy}')
print(classification_report(y_test, y_pred))


/Users/Ana/neuefische/Formula-1-Project/.venv/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/Users/Ana/neuefische/Formula-1-Project/.venv/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.or

Best hyperparameters: {'C': 0.1, 'penalty': 'l2'}
Accuracy: 0.661437908496732
              precision    recall  f1-score   support

           0       0.68      0.65      0.67       398
           1       0.64      0.67      0.66       367

    accuracy                           0.66       765
   macro avg       0.66      0.66      0.66       765
weighted avg       0.66      0.66      0.66       765



/Users/Ana/neuefische/Formula-1-Project/.venv/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/Users/Ana/neuefische/Formula-1-Project/.venv/lib/python3.11/site-packages/sklearn/model_selection/_validation.py:378: FitFailedWarning: 
30 fits failed out of a total of 60.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
-----------------------------------------------------------

The hyperparameter tuning and regularization helped slightly improve the model's performance compared to the previous Logistic Regression model, but the improvement is not substantial. The model still has similar precision, recall, and F1-score for both classes (0 and 1), and the overall accuracy is around 66.14%.



##### Decision Trees

In [61]:
from sklearn.tree import DecisionTreeClassifier


# Step 1: Data Preprocessing
X = balanced_df.drop(columns=['Class'])  # Features
y = balanced_df['Class']  # Target variable

# Step 2: Data Splitting
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Step 3: Create the Decision Tree classifier
tree_model = DecisionTreeClassifier(random_state=42)

# Step 4: Train the model on the training data
tree_model.fit(X_train, y_train)

# Step 5: Make predictions on the test data
y_pred = tree_model.predict(X_test)

# Step 6: Evaluate the model's performance
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy}')
print(classification_report(y_test, y_pred))


Accuracy: 0.9764705882352941
              precision    recall  f1-score   support

           0       1.00      0.95      0.98       398
           1       0.95      1.00      0.98       367

    accuracy                           0.98       765
   macro avg       0.98      0.98      0.98       765
weighted avg       0.98      0.98      0.98       765



These results indicate that the Decision Tree classifier is performing exceptionally well for our classification task. It has a very high accuracy, precision, recall, and F1-Score. The model is effectively identifying drivers who finish in the top 3 positions (class 1) and those who don't (class 0).

It's important to note that Decision Trees can sometimes overfit the training data, so we might want to consider techniques like pruning or using an ensemble of Decision Trees (Random Forest) to further improve the model's generalization and robustness.

Let's apply pruning to our Decision Tree model:

In [62]:
# Step 1: Data Preprocessing
X = balanced_df.drop(columns=['Class'])  # Features
y = balanced_df['Class']  # Target variable

# Step 2: Data Splitting
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Step 3: Create the Decision Tree classifier with Cost-Complexity Pruning
tree_model = DecisionTreeClassifier(random_state=42, ccp_alpha=0.01)  # Adjust the ccp_alpha value

# Step 4: Train the model on the training data
tree_model.fit(X_train, y_train)

# Step 5: Make predictions on the test data
y_pred = tree_model.predict(X_test)

# Step 6: Evaluate the model's performance
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy}')
print(classification_report(y_test, y_pred))


Accuracy: 0.8627450980392157
              precision    recall  f1-score   support

           0       0.88      0.86      0.87       398
           1       0.85      0.87      0.86       367

    accuracy                           0.86       765
   macro avg       0.86      0.86      0.86       765
weighted avg       0.86      0.86      0.86       765



Cost-Complexity Pruning has helped control the complexity of the Decision Tree and mitigate overfitting while maintaining good predictive performance. The model's accuracy is now around 86.27%, and it has balanced precision, recall, and F1-Scores for both classes (0 and 1).

This result suggests that the pruned Decision Tree is a suitable model for our classification task, providing a good balance between accuracy and generalization.

Here's a quick comparison:

KNN (k=3):

Accuracy: Approximately 89.93%
Balanced precision, recall, and F1-Scores for both classes.
High overall accuracy.

Pruned Decision Tree (with Cost-Complexity Pruning):

Accuracy: Approximately 86.27%
Balanced precision, recall, and F1-Scores for both classes.
Good accuracy and generalization, but slightly lower than KNN.

Based on accuracy alone, the KNN model outperforms the Decision Tree model. 








##### Random Forest

Let's try training a Random Forest classifier for our classification task. Random Forest is an ensemble learning method that combines multiple decision trees to improve predictive performance and reduce overfitting. 

In [63]:
from sklearn.ensemble import RandomForestClassifier

# Step 1: Data Preprocessing
X = balanced_df.drop(columns=['Class'])  # Features
y = balanced_df['Class']  # Target variable

# Step 2: Data Splitting
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Step 3: Create the Random Forest classifier
rf_model = RandomForestClassifier(random_state=42)

# Step 4: Train the model on the training data
rf_model.fit(X_train, y_train)

# Step 5: Make predictions on the test data
y_pred = rf_model.predict(X_test)

# Step 6: Evaluate the model's performance
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy}')
print(classification_report(y_test, y_pred))


Accuracy: 0.9764705882352941
              precision    recall  f1-score   support

           0       1.00      0.95      0.98       398
           1       0.95      1.00      0.98       367

    accuracy                           0.98       765
   macro avg       0.98      0.98      0.98       765
weighted avg       0.98      0.98      0.98       765



Let's perform cross-validation to assess the model's performance more robustly. Cross-validation is a technique that helps estimate how well a model will generalize to unseen data. In scikit-learn, we can use cross_val_score to perform cross-validation with our Random Forest model.

In [64]:
from sklearn.model_selection import cross_val_score

# Step 1: Data Preprocessing
X = balanced_df.drop(columns=['Class'])  # Features
y = balanced_df['Class']  # Target variable

# Step 2: Create the Random Forest classifier
rf_model = RandomForestClassifier(random_state=42)

# Step 3: Perform 5-fold cross-validation
# Adjust the number of folds (cv parameter) as needed
scores = cross_val_score(rf_model, X, y, cv=5, scoring='accuracy')

# Step 4: Print cross-validation results
print("Cross-Validation Scores:", scores)
print("Mean Accuracy:", scores.mean())


Cross-Validation Scores: [0.98300654 0.97254902 0.96727749 0.97774869 1.        ]
Mean Accuracy: 0.9801163467132052


These results are very promising and indicate that our Random Forest model is performing consistently well across different subsets of our data. The high mean accuracy of approximately 98.01% suggests that the model generalizes effectively and is likely to perform well on unseen data.

With such high cross-validation scores, we can have confidence in the robustness of our model. It's always a good practice to validate our model using cross-validation to ensure that it's not overfitting to the training data.

##### Confusion Matrix

In [65]:
# Create and fit the Random Forest classifier
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_train, y_train)

# Make predictions on the test data
y_pred = rf_model.predict(X_test)

# Calculate the confusion matrix
confusion = confusion_matrix(y_test, y_pred)

# Print the confusion matrix
print("Confusion Matrix:")
print(confusion)


Confusion Matrix:
[[380  18]
 [  0 367]]


- True Positives (TP): 367
- True Negatives (TN): 380
- False Positives (FP): 18
- False Negatives (FN): 0

TP (True Positives): The number of instances correctly predicted as belonging to the positive class (in our case, drivers ending a race in the top 3 positions).
TN (True Negatives): The number of instances correctly predicted as belonging to the negative class (drivers not ending in the top 3 positions).
FP (False Positives): The number of instances incorrectly predicted as belonging to the positive class when they actually belong to the negative class. In our case, it means predicting a driver to finish in the top 3 when they didn't.
FN (False Negatives): The number of instances incorrectly predicted as belonging to the negative class when they actually belong to the positive class. In our case, it means failing to predict a driver finishing in the top 3 when they did.


Our Random Forest model is performing remarkably well with no false negatives (FN), which means it correctly predicts all instances where drivers finish in the top 3 positions. The false positives (FP) are relatively low at 18, indicating a high level of precision.

This confusion matrix confirms that our Random Forest model is achieving an excellent level of accuracy and precision in predicting whether a driver ends a race in the top 3 positions. It appears to be a very strong model for our classification task.

#### Feature Importance

Feature importance indicates which features (variables) had the most influence on the model's predictions. Random Forest provides a straightforward way to calculate feature importance. 

In [66]:
importances = rf_model.feature_importances_

# Assuming you have a DataFrame with your feature names, you can associate them with their importance scores
feature_names = X.columns  # Replace X with your feature data

# Create a DataFrame to display feature names and their importance scores
feature_importance_df = pd.DataFrame({'Feature': feature_names, 'Importance': importances})

# Sort the features by importance in descending order
feature_importance_df = feature_importance_df.sort_values(by='Importance', ascending=False)

# Print or visualize the feature importance
print(feature_importance_df)


                      Feature  Importance
2                        grid    0.385617
0                    driverId    0.140514
26  pit_stop_1_milliseconds_x    0.046756
1                      raceId    0.042131
25           pit_stop_1_lap_x    0.039379
13                  Soft Used    0.031723
10                 Medium New    0.030142
28  pit_stop_2_milliseconds_x    0.029053
12                   Soft New    0.026297
27           pit_stop_2_lap_x    0.025663
15             Supersoft Used    0.022696
3                      year_x    0.021330
19                  circuitId    0.021301
30  pit_stop_3_milliseconds_x    0.019901
5                    Hard New    0.019041
17             Ultrasoft Used    0.015085
29           pit_stop_3_lap_x    0.014808
11                Medium Used    0.014089
14              Supersoft New    0.008968
8            Intermediate New    0.008036
18                    Wet New    0.005717
23               weather_warm    0.005361
20             weather_cloudy    0

As we can see, the grid feature has the highest importance score, indicating that it is the most influential feature for making predictions. The importance scores for other features vary, with some being more important than others. 